### PR TITLE
[SYCL] No nullptr program_impl

### PR DESCRIPTION
**Pull request by cperkinsintel**
The `kernel_impl` has been storing a `program_impl` pointer.  But none of the users of the `kernel_impl` need that, they just need the `PIProgram` .  Conversely, some of the callers constructing `kernel_impl` don't have a `program_impl` ptr and so we were using nullptr, which leads to crashes in some simple use cases.   

Replacing the `program_impl` ptr with a `PIProgram` member var instead and ensuring that all callers provide one, with the least amount of change to the `kernel_impl` interface. Simple test added as well - it crashes without this PR.

### CODE REVIEW

**Review by EwanC — APPROVED** [review]
`graph_impl.cpp` change LGTM

File: sycl/source/detail/kernel_impl.hpp [hunk 1]
```diff
@@ -31,6 +31,7 @@ class kernel_bundle_impl;
 
 using ContextImplPtr = std::shared_ptr<context_impl>;
 using ProgramImplPtr = std::shared_ptr<program_impl>;
+using ProgramPIPtr = sycl::detail::pi::PiProgram;
```
Comment (sergey-semenov) [inline sycl/source/detail/kernel_impl.hpp:34]: Nitpick: can we do this instead? Seems a bit more straightforward.
```suggestion
using sycl::detail::pi::PiProgram;
```

**Review by sergey-semenov — APPROVED** [review]
LGTM

**Comment by bader** [comment]
> @rdeodhar - this PR is to be the "real" fix, intended to replace the temporary workaround I gave you earlier ( #13148 )

@cperkinsintel, I think it's better to put messages like this in PR comments rather than PR description. I'm not sure if it's useful to have it as part of the git commit log message.

**Pull request merged**